### ISSUE TITLE
SetEncryptionSecret: State for: com.google.android.datatransport.events-journal not correct

### ISSUE THREAD

**Issue by t4tapas** [capacitor-community/sqlite#141]
**Describe the bug**
This is a new Ionic 5 app. So there is no existing database with secrets. I am trying to set the secure stored secret before dabase is created. 
I am calling method `setEncryptionSecret` before I am calling `createConnection` with `no-encryption`.

```
  private async setEncryptionSecret(passphrase: string) {
    if (this.sqlite != null) {
      try {
        const isSet = await this.sqlite.isSecretStored();
        if (!isSet.result) {
          await this.sqlite.setEncryptionSecret(passphrase);
        }
      } catch (error) {
         console.error(error)
      }
    }
  }
```

In android, I am getting the following error.

`SetEncryptionSecret: State for: com.google.android.datatransport.events-journal not correct`

IOS is working fine.

Update: 
The above error is coming on android. 8.0.0

I am getting the following new error in android 11

`SetEncryptionSecret: State for: androidx.work.workdb-wal not correct `

**Comment by jepiqueau** [comment]
@t4tapas can you share your gradle files and variables

**Comment by t4tapas** [comment]
@jepiqueau - following are the files

**variables.gradle**

```
ext {
    minSdkVersion = 23
    compileSdkVersion = 30
    targetSdkVersion = 30
    androidxActivityVersion = '1.2.0'
    androidxAppCompatVersion = '1.2.0'
    androidxCoordinatorLayoutVersion = '1.1.0'
    androidxCoreVersion = '1.3.2'
    androidxFragmentVersion = '1.3.0'
    androidxMaterialVersion = '1.3.0'
    androidxBrowserVersion = '1.3.0'
    androidxExifInterfaceVersion = '1.3.2'
    firebaseMessagingVersion = '21.0.1'
    junitVersion = '4.13.1'
    androidxJunitVersion = '1.1.2'
    androidxEspressoCoreVersion = '3.3.0'
    cordovaAndroidVersion =  '7.0.0'
    firebaseCrashlyticsVersion = '18.1.0'
}
```
**build.gradle** android level
```
buildscript {
    
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:4.2.0'
        classpath 'com.google.gms:google-services:4.3.3'
        classpath 'com.google.firebase:firebase-crashlytics-gradle:2.7.1'
    }
}

apply from: "variables.gradle"

allprojects {
    repositories {
        google()
        jcenter()
    }
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```
**build.gradle** android/app level
```

apply plugin: 'com.android.application'

android {
    compileSdkVersion rootProject.ext.compileSdkVersion
    defaultConfig {
        applicationId "com.test.testapp"
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

repositories {
    flatDir{
        dirs '../capacitor-cordova-android-plugins/src/main/libs', 'libs'
    }
}

dependencies {
    implementation fileTree(include: ['*.jar'], dir: 'libs')
    implementation "androidx.appcompat:appcompat:$androidxAppCompatVersion"
    implementation project(':capacitor-android')
    testImplementation "junit:junit:$junitVersion"
    androidTestImplementation "androidx.test.ext:junit:$androidxJunitVersion"
    androidTestImplementation "androidx.test.espresso:espresso-core:$androidxEspressoCoreVersion"
    implementation project(':capacitor-cordova-android-plugins')
    implementation 'com.google.firebase:firebase-crashlytics:18.1.0'
    implementation 'com.google.firebase:firebase-invites:17.0.0'
    implementation 'com.google.firebase:firebase-dynamic-links:20.1.0'
    implementation 'com.google.firebase:firebase-analytics:19.0.0'
}

apply from: 'capacitor.build.gradle'

try {
    def servicesJSON = file('google-services.json')
    if (servicesJSON.text) {
        apply plugin: 'com.google.gms.google-services'
    }
} catch(Exception e) {
    logger.warn("google-services.json not found, google-services plugin not applied. Push Notifications won't work")
}

 apply plugin: 'com.google.firebase.crashlytics'
```

Also, all the errors are the names of databases created when the app is installed.
![Screen Shot 2021-07-04 at 12 23 06 PM](https://user-images.githubusercontent.com/33161726/124376256-ba769880-dcc3-11eb-9ca7-380beb19bfe4.png)

Event after this error is raised, the database is getting created and it's encrypted with the provided passphrase.

**Comment by jepiqueau** [comment]
@t4tapas the gradle files seems correct. As I 1 currently travelling can you share on GitHub a app demonstrating the error so I could just clone it quickly as I do not really understand what is happening I never got that errror

**Comment by t4tapas** [comment]
@jepiqueau - that will take some time for me to create a similar app and share. 
I tracked the error to the `UtilsSecret.java` file in the plugin
The following `setEncryptionSecret()` method is fetching the state of all databases with method `getDatabaseState()` and if `state == DOES_NOT_EXIST || state == UNKNOWN` , its throwing the error with the database name.
 ```
    /**
     * SetEncryptionSecret
     * @param passphrase
     * @throws Exception
     */
    public void setEncryptionSecret(String passphrase) throws Exception {
        try {
            if (TextUtils.isEmpty(passphrase)) {
                String msg = "passphrase must not be empty";
                throw new Exception(msg);
            }
            // Store encrypted passphrase in sharedPreferences
            setPassphrase(passphrase);

            // Get the list of databases
            String[] dbList = uFile.getListOfFiles(context);
            if (dbList.length > 0) {
                for (String dbName : dbList) {
                    File file = context.getDatabasePath(dbName);

                    UtilsSQLCipher.State state = uCipher.getDatabaseState(context, file, sharedPreferences, globVar);
                    // change password if encrypted with globVar.secret
                    if (state == ENCRYPTED_GLOBAL_SECRET) {
                        uCipher.changePassword(context, file, globVar.secret, passphrase);
                    } else if (state == DOES_NOT_EXIST || state == UNKNOWN) {
                        String msg = "State for: " + dbName + " not correct";
                        throw new Exception(msg);
                    }
                }
            }
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }
```
Please let me know if this helps to figure out what might be wrong.

**Comment by jepiqueau** [comment]
@t4tapas Can you explain me what are the androidx.work.workdb-wal database How it has been created? Same for com.google.android.datatransport.events-journal

**Comment by jepiqueau** [comment]
@t4tapas this is fixed in 3.0.0

**Comment by t4tapas** [comment]
@jepiqueau - Validated. No more error. Thank you!!!
Closing this.
